HTMLImageElement and FormAssociatedElement should share more code

rniwa opened the pull request:
#### 5780eeea65cd07eaeb33633e80b07c8c2765d2f7
<pre>
HTMLImageElement and FormAssociatedElement should share more code
<a href="https://bugs.webkit.org/show_bug.cgi?id=247936">https://bugs.webkit.org/show_bug.cgi?id=247936</a>

Reviewed by Aditya Keerthi.

This patch makes HTMLImageElement and FormListedElement share more code in FormAssociatedElement.
m_form and m_formSetByParser are moved from those two classes to FormAssociatedElement along with
some logic to update the associated form.

* Source/WebCore/dom/Node.h:
* Source/WebCore/html/FormAssociatedElement.h:
(WebCore::FormAssociatedElement::~FormAssociatedElement): Moved the release assertion from
FormListedElement::~FormListedElement.
(WebCore::FormAssociatedElement::formWillBeDestroyed): Added.
(WebCore::FormAssociatedElement::form const): Added.
(WebCore::FormAssociatedElement::FormAssociatedElement): Added.
(WebCore::FormAssociatedElement::setForm): Added.
(WebCore::FormAssociatedElement::setFormInternal): Added.
(WebCore::FormAssociatedElement::elementInsertedIntoAncestor): Added.
(WebCore::FormAssociatedElement::elementRemovedFromAncestor): Added.

* Source/WebCore/html/FormListedElement.cpp:
(WebCore::FormListedElement::FormListedElement):
(WebCore::FormListedElement::~FormListedElement): Deleted.
(WebCore::FormListedElement::elementInsertedIntoAncestor): Renamed from insertedIntoAncestor.
Also takes the element and InsertionType as arguments.
(WebCore::FormListedElement::elementRemovedFromAncestor): Ditto from removedFromAncestor.
(WebCore::FormListedElement::form const): Deleted.
(WebCore::FormListedElement::formOwnerRemovedFromTree):
(WebCore::FormListedElement::setFormInternal): Extracted from the old setForm, which has been
moved to FormAssociatedElement.
(WebCore::FormListedElement::formWillBeDestroyed):
(WebCore::FormListedElement::resetFormOwner):
(WebCore::FormListedElement::formAttributeChanged):
(WebCore::FormListedElement::setForm): Deleted.

* Source/WebCore/html/FormListedElement.h:
(WebCore::FormListedElement::clearForm):

* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::insertedIntoAncestor):
(WebCore::HTMLFormControlElement::removedFromAncestor):

* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::~HTMLFormElement): Call formWillBeDestroyed instead of directly
updating HTMLImageElement&apos;s private member variable.

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::HTMLImageElement):
(WebCore::HTMLImageElement::form const): Deleted.
(WebCore::HTMLImageElement::resetFormOwner): Added.
(WebCore::HTMLImageElement::setFormInternal): Renamed from setForm.
(WebCore::HTMLImageElement::insertedIntoAncestor):
(WebCore::HTMLImageElement::removedFromAncestor):

* Source/WebCore/html/HTMLImageElement.h:

* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::HTMLObjectElement::insertedIntoAncestor):
(WebCore::HTMLObjectElement::removedFromAncestor):

* Source/WebCore/loader/FormState.cpp:

* Source/WebCore/loader/FormState.h:

Canonical link: <a href="https://commits.webkit.org/256697@main">https://commits.webkit.org/256697@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f33f2bf2f693a37b1a4fbc1e9556ea75eb3d9f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96532 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5785 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29603 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106059 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166405 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100514 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5970 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34523 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88904 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102775 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102205 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4453 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83132 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31417 "Found 1 new test failure: imported/w3c/web-platform-tests/fetch/stale-while-revalidate/stale-image.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86295 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88182 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40257 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37922 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21069 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4188 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43620 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2223 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1003 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40342 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Node.h:
```diff
@@ -267,7 +267,7 @@ class Node : public EventTarget {
     void setParentNode(ContainerNode*);
     Node& rootNode() const;
-    Node& traverseToRootNode() const;
+    WEBCORE_EXPORT Node& traverseToRootNode() const;
```
Why is `WEBCORE_EXPORT` necessary here?

rniwa replied:
I was getting this error without it:

```
  "WebCore::Node::traverseToRootNode() const", referenced from:
      WebCore::FormAssociatedElement::elementRemovedFromAncestor(WebCore::Element&, WebCore::Node::RemovalType) in Internals-ae7a10964258fecb6b8e55787c2964fe.o
      WebCore::FormAssociatedElement::elementRemovedFromAncestor(WebCore::Element&, WebCore::Node::RemovalType) in JSInternals.o
```

cdumez commented on Source/WebCore/html/FormAssociatedElement.h:
```diff
@@ -20,24 +20,83 @@
+    virtual void elementRemovedFromAncestor(Element&, Node::RemovalType);
+
+protected:
+    FormAssociatedElement(HTMLFormElement*);
```
explicit?

rniwa replied:
Will fix

cdumez commented on Source/WebCore/html/FormAssociatedElement.h:
```diff
@@ -20,24 +20,83 @@
+    virtual void setFormInternal(HTMLFormElement*);
+
 private:
+    WeakPtr<HTMLFormElement, WeakPtrImplWithEventTargetData> m_form;
```
Data members usually come last, it feels weird to have private functions *after* the data members.

rniwa replied:
Sure, will fix.

rniwa committed:
HTMLImageElement and FormAssociatedElement should share more code
https://bugs.webkit.org/show_bug.cgi?id=247936

Reviewed by Aditya Keerthi.

This patch makes HTMLImageElement and FormListedElement share more code in FormAssociatedElement.
m_form and m_formSetByParser are moved from those two classes to FormAssociatedElement along with
some logic to update the associated form.

* Source/WebCore/dom/Node.h:
* Source/WebCore/html/FormAssociatedElement.h:
(WebCore::FormAssociatedElement::~FormAssociatedElement): Moved the release assertion from
FormListedElement::~FormListedElement.
(WebCore::FormAssociatedElement::formWillBeDestroyed): Added.
(WebCore::FormAssociatedElement::form const): Added.
(WebCore::FormAssociatedElement::FormAssociatedElement): Added.
(WebCore::FormAssociatedElement::setForm): Added.
(WebCore::FormAssociatedElement::setFormInternal): Added.
(WebCore::FormAssociatedElement::elementInsertedIntoAncestor): Added.
(WebCore::FormAssociatedElement::elementRemovedFromAncestor): Added.

* Source/WebCore/html/FormListedElement.cpp:
(WebCore::FormListedElement::FormListedElement):
(WebCore::FormListedElement::~FormListedElement): Deleted.
(WebCore::FormListedElement::elementInsertedIntoAncestor): Renamed from insertedIntoAncestor.
Also takes the element and InsertionType as arguments.
(WebCore::FormListedElement::elementRemovedFromAncestor): Ditto from removedFromAncestor.
(WebCore::FormListedElement::form const): Deleted.
(WebCore::FormListedElement::formOwnerRemovedFromTree):
(WebCore::FormListedElement::setFormInternal): Extracted from the old setForm, which has been
moved to FormAssociatedElement.
(WebCore::FormListedElement::formWillBeDestroyed):
(WebCore::FormListedElement::resetFormOwner):
(WebCore::FormListedElement::formAttributeChanged):
(WebCore::FormListedElement::setForm): Deleted.

* Source/WebCore/html/FormListedElement.h:
(WebCore::FormListedElement::clearForm):

* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::insertedIntoAncestor):
(WebCore::HTMLFormControlElement::removedFromAncestor):

* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::~HTMLFormElement): Call formWillBeDestroyed instead of directly
updating HTMLImageElement's private member variable.

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::HTMLImageElement):
(WebCore::HTMLImageElement::form const): Deleted.
(WebCore::HTMLImageElement::resetFormOwner): Added.
(WebCore::HTMLImageElement::setFormInternal): Renamed from setForm.
(WebCore::HTMLImageElement::insertedIntoAncestor):
(WebCore::HTMLImageElement::removedFromAncestor):

* Source/WebCore/html/HTMLImageElement.h:

* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::HTMLObjectElement::insertedIntoAncestor):
(WebCore::HTMLObjectElement::removedFromAncestor):

* Source/WebCore/loader/FormState.cpp:

* Source/WebCore/loader/FormState.h:

Canonical link: https://commits.webkit.org/256697@main
